### PR TITLE
Move class `torch.nn.Parameter` out of module `torch.nn`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128633
* __->__ #130327
* #130326
* #130532

Move code in `torch.nn.parameter` to `torch._parameter`. Then import members into `torch.nn.parameter`. The classes's `__module__` attribute is set and also a `__getattr__` function is added to ensure backward compatibility.

### Rationale

The class `Parameter` does not depend on any function/class in module `torch.nn`. It can be moved out of `torch.nn` without any modification on `parameter.py`. It is a core feature of PyTorch and used in other submodules in PyTorch (e.g., `torch.autograd`, `torch.distributed`). Moving it out of module `torch.nn` allows us to prune the dependence between `torch.nn.*` and other PyTorch submodules.

This change is used to break the circular dependency between `torch.autograd` (requires `torch.nn.paramter.Parameter`) and `torch.nn` (requires `torch.autograd.grad_mode.no_grad`).

https://github.com/pytorch/pytorch/blob/a5f816df18f49619cfb15ffcd3b74606a495c4e2/torch/csrc/autograd/init.cpp#L123-L131

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @mruberry @jbschlosser @walterddr @mikaylagawarecki @mrshenli @pritamdamania87 @zhaojuanmao @satgera @aazzolini @osalpekar @jiayisuse @chauhang

### CODE REVIEW

**Review by albanD — CHANGES_REQUESTED** [review]
I'm not a fan of doing such code move without them being backed by a general plan.
Such change are very very risky as our users rely (too much) on non-public API.

Will follow on the discussion with @malfet on the broader plan here. But I don't think we should do this one as-is.

**Comment by XuehaiPan** [comment]
We have done a similar thing once before for extracting `torch._jit_internal` from the `torch.jit` submodule.

https://github.com/pytorch/pytorch/blob/cb2bce98de5bba1c5415a08d7b17c608f6c77b62/torch/_jit_internal.py#L2-L6

**Comment by albanD** [comment]
Sure but I don't see why we do it here. Parameter() is 100% tied to the concept of Module. You could say the same that Module doesn't depend on nn and so could be moved out.

File: torch/_parameter.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
+"""
+The :mod:`torch.nn.parameter` module provides a kind of :class:`~torch.Tensor`
+that is to be considered a module parameter.
+
+The class definitions need to be here instead of inside ``torch/nn/`` so they
+can be used in other places in ``torch/`` (namely :mod:`torch.autograd`) without
+running into circular dependency problems
+"""
```
Comment (XuehaiPan) [inline torch/_parameter.py:8]: > Sure but I don't see why we do it here. Parameter() is 100% tied to the concept of Module. You could say the same that Module doesn't depend on nn and so could be moved out.

I have added a description of why we need this.

------

Also, recap for the PR description.

This change is used to break the circular dependency between `torch.autograd` (requires `torch.nn.paramter.Parameter`) and `torch.nn` (requires `torch.autograd.grad_mode.no_grad`).

https://github.com/pytorch/pytorch/blob/a5f816df18f49619cfb15ffcd3b74606a495c4e2/torch/csrc/autograd/init.cpp#L123-L131

Comment (albanD) [reply torch/_parameter.py:8]: I'm not sure what you mean by this. These things work today, so we don't actually have a problematic dependency cycle?

**Pull request closed without merging**